src/wayland/selection/primary_selection/mod.rs

```diff
@@ -94,21 +94,32 @@ pub trait PrimarySelectionHandler: Sized + SeatHandler + SelectionHandler {
 }
 
 /// State of the primary selection.
-#[derive(Debug)]
 pub struct PrimarySelectionState {
     manager_global: GlobalId,
+    pub(super) filter: Arc<Box<dyn for<'c> Fn(&'c Client) -> bool + Send + Sync>>,
+}
+
+impl std::fmt::Debug for PrimarySelectionState {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("PrimarySelectionState").field("manager_global", &self.manager_global).finish()
+    }
 }
 
 impl PrimarySelectionState {
     /// Register new [`PrimaryDeviceManager`] global
-    pub fn new<D>(display: &DisplayHandle) -> Self
+    pub fn new<D, F>(display: &DisplayHandle, filter: F) -> Self
```

Comment:
No need to brake and complicate the API for such an esoteric use case. Let's just add another function `with_filter` like so: https://github.com/Smithay/smithay/blob/953959e6069b3e14dba96fdaa46c65990c21d5c9/src/wayland/shell/xdg/decoration.rs#L117-L155